### PR TITLE
Allow margins to collapse & refactor block toolbar

### PR DESCRIPTION
**Pull request by jasmussen**
The primary purpose of this PR is to allow margins to collapse. 

When two margins meet, and no borders or paddings are present, they collapse. I.e. a `p` with a 20px margin next to a similar `p` will have only 20px margin between them, not 40px, because the two siblings collapse to whichever margin is the largest. 

By allowing margins on blocks to collapse, we can make it easier to theme the editor to look like the frontend, by making margins behave the same.

This PR also refactors how the block toolbar is positioned. Because we use `position: sticky;`, the toolbar is part of the flow of the block itself. In this PR it uses `translate` to pull it _out of the flow_, meaning the block metrics do not have to account for the presence of the block toolbar or not. This is a nice refactor on its own, but it also makes the positioning more resilient and predictable.

## Testing

Because this is a rather large change to a visible part of the editor, it needs a good amount of testing. The gold standard is that **the editor in this branch should look no different than the editor in master**.

So please test on all breakpoints, test the frontend and backend, test with fixed toolbar and contextual toolbar:

- Test that the block toolbar that looks right on all blocks, including blocks like the spacer that do not have a block toolbar.
- Test all blocks they look right, margin-wise.
- Test in nested contexts blocks, test child blocks. 
- Test with editor styles. 

It's okay that there are a few regressions with existing editor styles, the benefits outweigh any fixes that need to be made.

### CODE REVIEW

**Comment by jasmussen** [comment]
This has been rebased, and I feel like it's in solid shape for wider testing, and I think we should get it in 3.2. Aside from the code cleanliness and benefits this has for editor-themers, the way the block outlines and toolbar are painted now are more resilient. Because it's hard to explain, here are two GIFs. 

Here, I am adjusting the margins on the _block content itself_, in master, where margins do not collapse:

without

Note how the block toolbar does not correctly position itself according to the adjusted margins. 

Here's from the branch that allows collapsing margins:

with

Note how changing just one margin on the block content doesn't do anything — because they collapse so the remaining margin is still there. Also notice how when the other margin is changed, the outlines and block toolbar correctly position themselves.

File: editor/components/block-list/style.scss
Comment (aduth) [inline editor/components/block-list/style.scss]: Mixed whitespace. Preceding space should be removed.

**Review by aduth — COMMENTED** [review]
Block toolbar is very misplaced in IE11

image

File: editor/components/block-list/style.scss
Comment (jasmussen) [reply editor/components/block-list/style.scss]: Nice catch, thank you.

**Comment by jasmussen** [comment]
Addressed feedback. Fixed the whitespace (nice catch, it shows up as a dot in VS code, I should've seen it but somehow missed it), and made it work in IE again. I hadn't properly rebased for Brandon's fix earlier, but now it works:

edge

ie

Noting also for others as well as myself, that `@supports( position: sticky ) { ... }` seems like a great way to target IE11 but not Edge, even for issues outside of position sticky. For example I might want to refactor the gallery hack that I made, to make things work in IE11, to use the above instead.

**Comment by jasmussen** [comment]
Rebased, and squashed to keep future rebases simpler. In a `fixup`'ed commit, https://github.com/WordPress/gutenberg/pull/7365/files#diff-00ebf096f7c483426f2a78d70d9cbd22R1003, I included a fix for a horizontal scrollbar regression introduced in #7225.

**Comment by jasmussen** [comment]
Unsure why Travis fails here. The tests pass.

**Comment by tofumatt** [comment]
That has been intermittently failing a lot. I've restarted it. 😢

**Review by karmatosed — APPROVED** [review]
Design side approving and did some testing. I would like to have more testing just to be super sure.

**Comment by jorgefilipecosta** [comment]
Hi @jasmussen, thank you for this changes. I confirm it solves a regression introduced during the refactoring to the block hover area.
I noticed a regression when clicking on the toolbar even if we move the mouse one pixel we trigger multi-selection. Before that was not the case. I'm not certain what is causing this problem maybe solve z-index change.
jul-03-2018 16-09-54

**Comment by tofumatt** [comment]
FYI fixed the merge conflict here (via `git rebase`, so anyone working on this branch'll need to pull this new version).

**Review by tofumatt — COMMENTED** [review]
This works for me from my testing, but I noticed the same regression @jorgefilipecosta did.

I found the cause… I'll push a commit fix for it.

**Comment by tofumatt** [comment]
I pushed a fix (https://github.com/WordPress/gutenberg/pull/7365/commits/cb8aae8de4c21993f8b4b879694425637fc6771e) for the regression with mouse movement on `mouseDown` @jorgefilipecosta mentioned and rebased this against master as there have been a few conflicts.

File: editor/components/block-list/block.js
Comment (jasmussen) [reply editor/components/block-list/block.js]: This move was initially the key to unlocking collapsing margins. So I assumed that aspect would break when you moved this back. 

However later on I added https://github.com/WordPress/gutenberg/pull/7365/commits/3c0afac08b5cdbac268793329ffed27e9674d604#diff-00ebf096f7c483426f2a78d70d9cbd22R904, which as it turns out is sufficient to enable collapsing margins, even without being relocated.

However all the in-between styles were written to assume the position of the nested toolbar where it was before this move, so there were a ton of visual regressions as a result of this move, like toolbars not being in the right place, and so on. But I'm fixing those now, and in the end I think it was good to move this back. Thanks for all the help.

**Comment by jasmussen** [comment]
Okay, so I fixed a _ton_ of issues that propped up after the columns merge and the fix to the mouse movement regression. So many that I think this needs another deep dive into the code. Can you take a look, @tofumatt? Thanks. 

I left comments for each fix in each commit.

Once again tests pass for me with `npm run test` but Travis fails. Not sure why.

Also, in master and this branch, every image is broken for me, even after an `npm install`, I'm getting this error:

```
react-dom.min.82e21c65.js:110 TypeError: this.props.forwardedRef is not a function
    at Wrapper.handleRef (index.js:58)
    at cg (react-dom.min.82e21c65.js:110)
    at dg (react-dom.min.82e21c65.js:110)
    at eg (react-dom.min.82e21c65.js:113)
    at jc (react-dom.min.82e21c65.js:131)
    at gc (react-dom.min.82e21c65.js:127)
    at vb (react-dom.min.82e21c65.js:126)
    at ub (react-dom.min.82e21c65.js:126)
    at zd (react-dom.min.82e21c65.js:124)
    at ra (react-dom.min.82e21c65.js:123)
```

But given that's also in master, I'm guessing it's unrelated to this branch.

File: editor/components/block-list/style.scss
Comment (tofumatt) [inline editor/components/block-list/style.scss]: Typo: "preceding". English is weird. I've pushed a fix 😄

**Comment by tofumatt** [comment]
`npm install` works for me so I'm guessing it's local to your install, and that failing travis test is… frequent and happening a lot. We'll need to look into it. 😢

**Comment by jasmussen** [comment]
To clarify, npm install is working fine for me. What isn't, is any image or embed. Try loading the demo content, I'm pretty sure you should see the same issue.

**Comment by tofumatt** [comment]
As mentioned in Slack, this seems unrelated but a legit breakage: https://wordpress.slack.com/archives/C02QB2JS7/p1530703711000116

**Review by tofumatt — APPROVED** [review]
<del>I like this a lot more, it feels more like a real, empty canvas (and more like the old editor) with the tighter margins.</del> I have no idea what I'm on about and I think that's an unrelated change. I see what this is doing now. All good. More coffee needed.

I think we need to sort out that image bug but as mentioned it's not related to this PR and seems to be happening in master.

🚢 … but maybe after we fix master just so we don't merge breaking builds 😉

**Comment by tofumatt** [comment]
As mentioned in Slack though: this is a great example of a PR that should include before/after screenshots. The differences are easy to spot visually but I found the lack of issue to make it unclear what I was testing for.

**Comment by jasmussen** [comment]
Thanks a bunch for the review. I'll wait for master to behave and then merge, possibly rebase first.

**Comment by jasmussen** [comment]
Rebased. Images work now, and the rest does too. 

How do we figure out if the travis failure is real? And otherwise, can we merge?

**Comment by gziolo** [comment]
> How do we figure out if the travis failure is real? And otherwise, can we merge?

It would be very surprising if CSS changes would make any test fail. I retriggered another build.

File: editor/components/block-list/style.scss
Comment (gziolo) [inline editor/components/block-list/style.scss]: We should remove this line before merging.

Comment (jasmussen) [reply editor/components/block-list/style.scss]: Fixed, thanks. I suppose this triggered another test? :)

File: test/e2e/specs/adding-blocks.test.js
Comment (aduth) [inline test/e2e/specs/adding-blocks.test.js]: This refactor may have made the test less effective at covering the expected behavior: https://github.com/WordPress/gutenberg/pull/11168#pullrequestreview-169297243

Comment (jasmussen) [reply test/e2e/specs/adding-blocks.test.js]: What's the best way to enhance this test again? CC: @gziolo who graciously helped me with this.

Comment (gziolo) [reply test/e2e/specs/adding-blocks.test.js]: Ooops, I wasn't aware of that :(

**Pull request merged**
